consensus/common_test.go

```diff
@@ -278,7 +280,8 @@ func newConsensusStateWithConfigAndBlockStore(thisConfig *cfg.Config, state sm.S
 	evpool := sm.MockEvidencePool{}
 
 	// Make ConsensusState
-	stateDB := dbm.NewMemDB()
+	stateDB := blockDB
+	sm.SaveState(stateDB, state) //for save height 1's validators info
```

Comment:
I don't love how we have to remember to do this everywhere. And again why does the stateDB have to be the same as blockDB?

Maybe we should have a `newStateDB(state)` function in this package that we could use in all these tests instead?

Reply:
It's because in reactor_test the func startConsensusNet
if css[i].state.LastBlockHeight == 0 { //simulate handle initChain in handshake
			sm.SaveState(css[i].blockExec.Db(), css[i].state)
		}
we should fetch the stateDB out, but I don't get a way, I have to add a public method in blockExecutor,  so that they have to be the same.

Reply:
not sure I follow. why can't we use `stateDB := dbm.NewMemDB()` here? Isn't the BlockExecutor supposed to use the stateDB?

Reply:
Before you suggest, I use startConsensusNet in TestSimulateValidatorsChange, which you said is too complicated and expensive just to simulate some blocks with validator change txs.
In that situation, the state must get saved in startConsensusNet, because later it will be used in replay test.
Now, I modifed TestSimulateValidatorsChange using no reactor, just start one ConsensusState, so I think deleting the SaveState code in startConsensusNet is OK. But think of the consistency of the code, we do saveState if we initChain in handshake, so startConsensusNet is to simulate that handshake, it is supposed to do exactly the same action.

Reply:
If startConsensusNet has to call SaveState, that is the problem I encountered, I have no way to fetch the stateDB in that test method, that is why I set blockExec.Db() and stateDB to be a same one.    Because I can fetch  blockExec.Db()  here.

Reply:
I see you have opened a new issue to optimize the simulate test case. Um, I think without consensus process is feasible. As you said, the test case is quite large now.   Maybe call several makeBlock() with some precommits and validator change txs will work.

Reply:
> I don't love how we have to remember to do this everywhere. And again why does the stateDB have to be the same as blockDB?
> 
> Maybe we should have a `newStateDB(state)` function in this package that we could use in all these tests instead?

I think this could work, we should ensure this function returns a singleton stateDB, that is if the state is passed in at the first time, it returns a new stateDB, else it returns a existed one.